### PR TITLE
Add support for ordinal formatting (and update CLDR)

### LINKED ISSUE #51: `selectordinal` support missing
I was attempting to format the following string and hit "Format 'selectordinal' could not be resolved":

> "{index, selectordinal, other {Slot #}}"

This is a toy example where in English I'd have "Slot 1", "Slot 2", etc - `string.Format` works fine, but the point would be to support other variants in a more complex string or in non-English cultures.

I believe technically, `selectordinal` is part of the MF 2.0 spec, I'm not sure if this lib is intended to target the "legacy" spec, MF 2.0, or a hybrid. 

Not expecting a magic fix here but wanted to track the issue - is this something that would be solved with a new formatter? I believe the data we need for ordinal rules exists in the CLDR repo along with plurals, but I haven't looked into what integration would take, yet.

### LINKED ISSUE #53: Built-in English pluralizer is inconsistent w/CLDR
See: https://github.com/jeffijoe/messageformat.net/blob/79da44e80b4bd1d7f3ad9ad7dd44e52a50e08d86/src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs#L299C29-L299C33

In `PluralFormatter` today we have a "default" pluralizer from before the CLDR days of 2021 - 

```cs
            n =>
            {
                if (n == 0)
                {
                    return "zero";
                }

                if (n == 1)
                {
                    return "one";
                }

                return "other";
            }
```

Unfortunately "zero" isn't actually a supported English form - the proper MessageFormat selector is `=0` for this scenario in English.

Then in [Pluralize](https://github.com/jeffijoe/messageformat.net/blob/79da44e80b4bd1d7f3ad9ad7dd44e52a50e08d86/src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs#L135), we use the builtin `Pluralizers` prior to checking the codegen'd CLDR stuff - so the built-in English takes priority and CLDR for English is never used.

It's an easy enough code fix but doing that would:

* Breaks several existing tests that rely on "zero" in English
* Breaks anyone using the library with a "zero" string in prod today

Not sure if the README should be more explicit about wanting to remove the default formatters, or make the breaking change behind a major version bump, or address in some other way.

---

The impact is that given the string `"{n, plural, zero {none} one {1} other {#}}"`, formatting `{ "n": 0 }` *should* return "0" in English instead of "none".

The correct MessageFormat English string for this scenario would be `"{n, plural, =0 {none} one {1} other {#}}"`

### PR DESCRIPTION
**Pull request by sfuqua**
Fixes #51 

This change adds support for the `selectordinal` function/formatter. Though I don't believe `selectordinal` is documented in the original ICU libraries, it seems to have ecosystem support and is a first class function in MF2, so it seems a good fit.

Thankfully the CLDR `ordinals.xml` has the same schema as `plurals.xml` so the parsing logic could be reused, though I had to update the data structures and generated code to support multiple plural "types".

1. Add `ordinals.xml` from CLDR v48.1 (and update `plurals.xml` simultaneously)
2. Add README to the data folder describing pedigree of the XML files
3. Add documentation to some of the codegen types and `PluralContext` to serve as touchstones
4. Update parsing infra to support two XML files and two plural "types" throughout the stack:
    a. Collections of `PluralRule` are replaced with a new data structure `PluralRuleSet` that serves as an index of locales + plural type
    b. There are now two different TryGetByLocale functions in the generated code, one for Cardinal (current behavior) and one for Ordinal (new/added behavior)
 5. Per discussion in this thread, removed default English pluralizers, which fixes #53 
 6. Update existing tests to explicitly exercise cardinal formatting
 7. Added some new ordinal test cases
 8. Added rudimentary support for LMDL "inheritance" rules, such that i.e. "en-CA" will match CLDR "en", and unknown or unmapped locales will always change to "root"; removed the library's default mapping to "en" as a result (because "root" exists)

### CODE REVIEW

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/PluralRuleSet.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                new PluralRuleKey(PluralType: pluralType, Locale: locale),
+                newRuleIndex
+            );
+        }
+    }
+
+    /// <summary>
+    /// Used to retrieve a specific <see cref="PluralRule"/>.
+    /// </summary>
+    public readonly record struct PluralRuleKey(string PluralType, string Locale)
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/PluralRuleSet.cs]: Wasn't able to use the shorthand for `readonly record struct` in this csproj due to the missing `IsExternalInit` class - might be due to the generator targeting only netstandard2.0?

The mirror type in the main MessageFormat project doesn't have the same issue 🤷 but manually making this one immutable wasn't too bad

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs [hunk 2]
```diff
@@ -29,18 +28,14 @@ public string GenerateClass()
...
+        for (var ruleIdx = 0; ruleIdx < uniqueRules.Count; ruleIdx++)
         {
-            var rule = _rules[ruleIdx];
-
+            var rule = uniqueRules[ruleIdx];
             var ruleGenerator = new RuleGenerator(rule);
 
-            foreach(var locale in rule.Locales)
-            {
-                WriteLine($"public static string Locale_{locale.ToUpper()}(PluralContext context) => Rule{ruleIdx}(context);");
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs:40]: These seemed unused - can add them back as `Locale_{type}_{locale}` or something if needed

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs [hunk 3]
```diff
@@ -287,10 +335,10 @@ internal string ReplaceNumberLiterals(string pluralized, double n)
     /// <summary>
     ///     Adds the standard pluralizers.
     /// </summary>
-    private void AddStandardPluralizers()
+    protected virtual void AddStandardPluralizers()
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs]: Oops, I think this was from a very early iteration where I subclassed `PluralFormatter` - can revert this to private

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/AST/Condition.cs [hunk 4]
```diff
@@ -3,6 +3,19 @@
 
 namespace Jeffijoe.MessageFormat.MetadataGenerator.Plural.Parsing.AST;
 
+/// <summary>
+/// Represents the 'condition' part of the <see href="https://unicode.org/reports/tr35/tr35-numbers.html#plural-rules-syntax">LDML grammar</see>.
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/AST/Condition.cs]: Realized looking over the change that my new docs are mostly inconsistent with how this project does xmldoc - need to indent everything one level

File: src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs [hunk 5]
```diff
@@ -65,20 +69,56 @@ public void Ru_PluralizerTests(double n, string expected)
     [InlineData(101, "days")]
     [InlineData(102, "days")]
     [InlineData(105, "days")]
-    public void En_PluralizerTests(double n, string expected)
+    public void En_Cardinal_PluralizerTests(double n, string expected)
     {
         var subject = new PluralFormatter();
+        subject.Pluralizers.Clear();
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs]: I noticed this test was originally exercising the "standard"/built-in (non-CLDR) rules for "en" - which gets exciting when you realize that CLDR English doesn't have a "zero" rule, but the built-in one does 😁

I left the default behavior alone in the rest of the project, but for this test suite I'm now clearing the defaults in each test.

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/PluralRuleSet.cs (hunk 1 above)
Comment (jeffijoe) [reply src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/PluralRuleSet.cs]: Feel free to install [PolySharp](https://github.com/Sergio0694/PolySharp) in the generator project.

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs (hunk 2 above)
Comment (jeffijoe) [reply src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs:40]: Maybe @kostya9 can confirm, but if truly unused, feel free to remove.

File: src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs (hunk 5 above)
Comment (jeffijoe) [reply src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs]: I responded to the new issue you created about this - I think it makes sense to fix it and ship a major release.

I couldn't tell you _why_ I added that `zero` back then, I probably didn't know about the CLDR and thought having `zero` was cute. 😅

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/AST/Condition.cs (hunk 4 above)
Comment (jeffijoe) [reply src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/Parsing/AST/Condition.cs]: Usually ReSharper formatter takes care of that.

I appreciate the additional XML comments by the way! 🙏

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralContext.cs [hunk 6]
```diff
@@ -60,26 +77,77 @@ private PluralContext(string number, double parsed)
...
+    /// <summary>
+    /// The count of visible fraction digits of <see cref="Number"/>, with trailing zeroes.
+    /// </summary>
+    /// <example>
+    /// 1.450 -> V = 3
+    /// </example>
     public int V { get; }
 
+    /// <summary>
+    /// The count of visible fraction digits of <see cref="Number"/>, without trailing zeroes.
```
Comment (jeffijoe) [inline src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralContext.cs]: These comments are incredibly useful, thanks again for adding them! 🙏

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs [hunk 7]
```diff
@@ -306,7 +354,34 @@ private void AddStandardPluralizers()
 
                 // ReSharper restore CompareOfFloatsByEqualityOperator
                 return "other";
-            });
+            }
+        );
+        this.Pluralizers.Add(
```
Comment (jeffijoe) [inline src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs]: As mentioned, these would go away if we make the change discussed in #53 right?

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralRuleKey.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Jeffijoe.MessageFormat.Formatting.Formatters
+{
```
Comment (jeffijoe) [inline src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralRuleKey.cs]: Let's use file-scoped namespaces.

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs (hunk 7 above)
Comment (sfuqua) [reply src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs]: Indeed

**Comment by sfuqua** [comment]
@jeffijoe I'm happy to tackle #53 (and update the README) in the same change. I figured as-is, this change could be a semver minor version bump for a backcompat new feature, but if you're happy to call it a major version change I'll go ahead and make the 'zero' change as well (which would just leave the open question about the codegen change). I can also do it in a separate PR - lmk if you have a preference.

I can also update the codegen to be backwards compatible by keeping the _LANG helpers and adding a new overload of `TryGetRuleByLocale` for ordinals, but I was pretty sure the generator project and what it spits out isn't part of the versioned API surface.

**Comment by jeffijoe** [comment]
The codegen is not part of the public API, so if everything works without the extra generated code, then we should leave it out.

Happy to make it a major bump now - I'd want to also tackle #41 in the same release since that is also a breaking change. Basically all that is for is making the default culture `InvariantCulture` as well as allowing passing a `CultureInfo` per invocation.

**Comment by sfuqua** [comment]
I was cleaning up `PluralRuleKey` a bit and realized I need to check some behavior about how input is matched against supported locales to ensure that e.g. "fr-FR" matches "fr".

Going to double check master with some new tests to confirm I didn't inadvertently change any behavior around fallback, and also look into how straightforward it is to canonicalize an input locale to a Unicode language ID and use the LDML matching algorithm to pick a rule.

**Comment by sfuqua** [comment]
Okay, I discovered a few other things that I'm interested in trying to contribute but that may be outside the scope of this PR through a reading of https://www.unicode.org/reports/tr35/tr35.html#Locale_Inheritance

1. There's no language inheritance in the lib today, so "fr-FR" _should_ match "fr", but does not (can confirm this in the existing tests by updating "ru" to "ru-RU"
2. The CLDR plural data has two examples of subtags ("pt_PT" and "kok_Latn") - per the spec, CLDR currently always uses "_" as the separator
3. Per the spec, -/_ are equivalent, so an input to the library of "pt-PT" should match "pt_PT"
> The "-" and "_" separators are treated as equivalent, although **"-" is preferred**.
4. The lookup of locales should be case-insensitive (alternatively, the input should be canonicalized)
5. The LDML inheritance/matching algo should be implemented, such that:

> Given a particular locale id "en_US_someVariant", the default search chain for a particular resource is the following.
> en_US_someVariant
> en_US
> en
> root

There is nuance in how to do search chaining for non-canonical names, however, given that almost all the plural data uses "base" language tags, I'm not sure there's value in going that far.

Notably, implementing support for `root` means *every input will always map to a default Pluralizer* based on CLDR rules, so we won't have to fallback to "en" anymore for missing languages.

I think my plan for this change is to -

1. Update the codegen to normalize CLDR `_` to BCP47 `-` in the Dictionary - this will hopefully make "pt-PT" work more often by default, and maintain a separate index of tags that include underscores to facilitate conversion in step 4.
2. Restructure the codegen to use `Dictionary<string, Dictionary<string, ContextPluralizer>>` keyed on CLDR-locale, and then on type (or vice versa), with case-insensitive comparison for the locale Dictionary. I'll keep using a `record struct` for the user-facing API, but not as a Dictionary key.
3. Update documentation to clarify the semantics of the `locale` string in a couple places.
4. Call `TryGetRuleByLocale` successively up to the root language before failing. If we detect an underscore, we can reference the index in step 1 to remap to a support locale (e.g., "pt_PT" -> "pt-PT" -> rule).
5. Document the edge cases in README
6. New tests

I think step 2 (the Dictionary change) should happen regardless as it's directly related to the feature in this PR, but the rest of the fallback stuff could wait til a separate change. I'll probably start on it in parallel and let me know if you have a preference on logistics (or if this should be tracked with a new issue).

**Comment by jeffijoe** [comment]
Don't want to bog you down with logistics, but if you feel like it would be simpler for both of us to merge this PR first, we can do that, let me know when it's ready and I'll merge it. Appreciate the help!

**Comment by sfuqua** [comment]
Okay, this ended up a little silly but I landed a compromise for locale inheritance in a new `LocaleHelper.cs`.

We try the original locale as-is (now case insensitive!), then we try stripping all subtags (so en-US-Foo will try "en"), then we finally try "root". 

This means - 
1. No more fallback to English (instead relying on "root" CLDR rules) for unknown languages - this is technically another "breaking" change, though depending on this behavior would've been odd
2. Most valid BCP 47 tags or CultureInfo names passed into this library should now properly match a CLDR rule when they would've hit English before
3. Built-in English pluralizers have been removed (can be added back in manually), and tests have been updated

This should work very well for almost all pluralization cases except for "pt-PT" with additional subtags (which would fall back to "pt", which is equivalent to "pt-BR").

We can get away with this because almost all of the plural CLDR locales are base languages.

I did end up substantially refactoring my previous codegen to key on locale first, and made Cardinal/Ordinal strongly typed instead of passing "cardinal" and "ordinal" strings everywhere.

File: src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs (hunk 5 above)
Comment (sfuqua) [reply src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs]: This change now fixes #53 in the latest iterations

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs (hunk 2 above)
Comment (sfuqua) [reply src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs:40]: Resolving thread for now

File: src/Jeffijoe.MessageFormat/Helpers/LocaleHelper.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    ///     which almost exclusively use languages without subtags.
+    /// </remarks>
+    /// <example>
+    ///     Given "language-Script-REGION", yields:
+    ///     - language-Script-REGION
+    ///     - language
+    ///     - root
+    /// </example>
+    /// <param name="locale">A BCP 47 locale tag</param>
+    public static IEnumerable<string> GetInheritanceChain(string locale)
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat/Helpers/LocaleHelper.cs:31]: this is the most substantial new code in the latest push - the rest of it is tweaking previous work (going from the `PluralLookupKey` I'd introduced in the first iteration, back to string locales but with explicit CLDR lookup delegates for each type)

the codegen and `PluralRuleSet` changed kind of substantially but not in very exciting ways

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs [hunk 10]
```diff
@@ -128,26 +191,42 @@ public string Format(string locale,
...
+        string? pluralForm = null;
+        if (customLookup.TryGetValue(locale, out var pluralizer))
         {
             pluralForm = pluralizer(context.Number);
         }
-        else if (PluralRulesMetadata.TryGetRuleByLocale(locale, out var contextPluralizer))
+        else
         {
-            pluralForm= contextPluralizer(context);
+            foreach (var candidate in LocaleHelper.GetInheritanceChain(locale))
```
Comment (sfuqua) [inline src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs:214]: This is where the new helper is invoked

File: src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs [hunk 11]
```diff
@@ -23,7 +39,7 @@ public class PluralFormatter : BaseFormatter, IFormatter
     public PluralFormatter()
     {
         this.Pluralizers = new Dictionary<string, Pluralizer>();
```
Comment (jeffijoe) [inline src/Jeffijoe.MessageFormat/Formatting/Formatters/PluralFormatter.cs]: Feel free to rename this to `CardinalPluralizers` for consistency since we're doing a major release anyway.

File: src/Jeffijoe.MessageFormat/MessageFormatter.cs
Comment (jeffijoe) [inline src/Jeffijoe.MessageFormat/MessageFormatter.cs]: Can rename this, too.

**Review by jeffijoe — COMMENTED** [review]
Since we're already doing a major for the next release, we can rename `Pluralizers` to `CardinalPluralizers` for consistency.

File: src/Jeffijoe.MessageFormat/MessageFormatter.cs
Comment (sfuqua) [reply src/Jeffijoe.MessageFormat/MessageFormatter.cs]: Done, and updated the README

File: src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs [hunk 12]
```diff
@@ -43,6 +47,8 @@ public void Uk_PluralizerTests(double n, string expected)
     public void Ru_PluralizerTests(double n, string expected)
     {
         var subject = new PluralFormatter();
+        subject.CardinalPluralizers.Clear();
```
Comment (jeffijoe) [inline src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs]: Do we still need these clears after making the change to default pluralizers?

Comment (sfuqua) [reply src/Jeffijoe.MessageFormat.Tests/MetadataGenerator/GeneratedPluralRulesTests.cs]: Nope!

**Review by jeffijoe — APPROVED** [review]

**Comment by jeffijoe** [comment]
Merged, thank you so much! Was there anything else you wanted to address before I cut the next major release?

**Comment by sfuqua** [comment]
> Merged, thank you so much! Was there anything else you wanted to address before I cut the next major release?

Nothing blocking on my end! Please feel free to tag me if anything strange surfaces with the changes.

Longer term I'm interested in a C#/.NET solution for [MessageFormat 2](https://messageformat.unicode.org/docs/quick-start/) which got ratified this year - it'd be a bit of a project but the package might actually be set up well for a dual parsing pipeline in the future leveraging all the same generated CLDR bindings.

The [lib](https://messageformat.github.io/modules/messageformat) I believe you originally used as a reference is now one of the reference implementations of the new syntax 😁

**Comment by jeffijoe** [comment]
Yes, it's been a long time, I didn't realize there was a new spec. 😅 

I'll probably take a stab at #41 before I cut a new release, but I can't promise when that will be.

EDIT: Took a quick look at MF2, that is definitely a big undertaking! But I like the syntax!

EDIT2: this is nuts

```
.input {$pronoun :string}
.input {$count :number}
.match $pronoun $count
he one   {{He has {$count} notification.}}
he *     {{He has {$count} notifications.}}
she one  {{She has {$count} notification.}}
she *    {{She has {$count} notifications.}}
* one    {{They have {$count} notification.}}
* *      {{They have {$count} notifications.}}
```

**Comment by jeffijoe** [comment]
This has now been [released](https://github.com/jeffijoe/messageformat.net/releases/tag/8.0.0)! Thanks again!

File: src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs (hunk 2 above)
Comment (kostya9) [reply src/Jeffijoe.MessageFormat.MetadataGenerator/Plural/SourceGeneration/PluralRulesMetadataGenerator.cs:40]: That was a long time ago 😄 If I remember correctly, I was not sure whether to use Rule-based or Locale-based functions, and finally decided on the rule-based ones, but didn't actually clean up the other ones. Looks good from my side
